File: 백준/Silver/10819. 차이를 최대로/차이를 최대로.py
n = int(input())
array = list(map(int,input().split()))
calc = []
visited = [0] * n
max_value = float('-inf')
def recursive(idx):
    global max_value
    # basis part
    if idx == n:
        value = 0
        # 연산 실행
        for i in range(n - 1):
            value += abs(array[calc[i]] - array[calc[i + 1]])
        max_value = max(max_value, value)
        return

    # inductive part
    for i in range(n):
        if not visited[i]:
            visited[i] = 1
            calc.append(i)
            recursive(idx + 1)
            visited[i] = 0
            calc.pop()

recursive(0)
print(max_value)

File: 백준/Silver/10819. 차이를 최대로/차이를 최대로.cc
#include <iostream>
#include <stdlib.h>
#include <limits.h>
#include <vector>
using namespace std;
vector<int> visited;
vector<int> calc;
vector<int> array;
int n;
int result = INT_MIN;

void recursive(int idx) {
    // basis part
    if (idx == n) {
        int res = 0;
        for (int i = 0; i < n - 1; i++) {
            res += abs(array[calc[i]] - array[calc[i + 1]]);
        }
        result = max(result, res);
    }
    
    // inductive part
    for (int i = 0; i < n; i++) {
        if (!visited[i]) {
            visited[i] = 1;
            calc.push_back(i);
            recursive(idx + 1);
            visited[i] = 0;
            calc.pop_back();
        }
    }
}

int main() {
    cin >> n;
    visited.resize(n, 0);
    array.resize(n);
    for (int i = 0; i < n; i++) {
        cin >> array[i];
    }
    recursive(0);
    cout << result;
}